staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,256 @@
+	abortedCh chan<- struct{}
+}
+
+// updateQueue is a queue of StorageVersion updates. Upon creation, a goroutine
+// is also created which keeps processing pending updates in the queue, until
+// the queue is closed.
+type updateQueue struct {
+	// q is the actual queue. A user can send StorageVersion updates to the
+	// queue. A goroutine runs in the background keeps processing the
+	// pending updates and doing the actual work, until the queue is closed.
+	q chan<- *update
+	// All the updates in the queue share the same context. Calling cancel
+	// aborts all the pending updates in the queue. This function is used
+	// when a CRD is deleted and we want to release all the associated
+	// resources (channel and goroutine). The caller should also close q to
+	// stop the background goroutine.
+	cancel context.CancelFunc
+}
+
+// manager implements the Manager interface.
+type manager struct {
+	// updateQueues holds a CRD UID to updateQueue map. Each CRD has its
+	// own queue of StorageVersion updates, and a goroutine which processes
+	// the pending updates in the queue. The manager sends update requests
+	// to the queue.
+	updateQueues map[types.UID]*updateQueue
+	// client is the client interface that manager uses to update
+	// StorageVersion objects.
+	client genericstorageversion.Client
+	// apiserverID is the ID of the apiserver that invokes this manager.
+	apiserverID string
+}
+
+// NewManager creates a CRD StorageVersion Manager.
+func NewManager(client genericstorageversion.Client, apiserverID string) Manager {
+	return &manager{
+		updateQueues: make(map[types.UID]*updateQueue),
+		client:       client,
+		apiserverID:  apiserverID,
+	}
+}
+
+// UpdateStorageVersionFor queues a StorageVesrion update for the given
+// CRD and returns immediately. Optionally, the caller may specify a
+// non-nil waitCh and/or a non-nil processedCh and a non-nil abortedCh.
+// A non-nil waitCh will block the StorageVersion update until waitCh is
+// closed.
+// The manager will close the non-nil processedCh if it finished
+// processing the StorageVersion update (note that the update can either
+// succeeded or failed), or close the non-nil abortedCh if it aborted
+// processing the update.
+func (m *manager) UpdateStorageVersionFor(crd *apiextensionsv1.CustomResourceDefinition,
+	waitCh <-chan struct{}, processedCh, abortedCh chan<- struct{}) {
+	q := m.getOrCreateUpdateQueue(crd.UID)
+	// When the channel buffer is full, writing to the channel becomes
+	// blocking. Here we give up updating storage version for this CRD and
+	// print a log, so that we can return immediately and not block the
+	// caller.
+	if len(q) == updateQueueBufferSize {
+		// TODO(roycaihw): use warning instead of info when StorageVersionAPI
+		// graduates to beta/GA
+		klog.V(2).Infof("Skipping the storage version update for CRD with UID %v due to the queue being full (queue size: %v).",
+			crd.UID, updateQueueBufferSize)
+		if processedCh != nil {
+			close(processedCh)
```

Comment:
does this case mean that another CRD state in time is queued already and will be used to update the storage version, but we still claim that it got processed? How do we guarantee consistency? Would rather expect to put just one update per UID into the queue, and replace the unprocessed update with a more up-to-date one, or even merge them.

Reply:
> one update per UID into the queue

In theory it's possible that a CRD's storage version changes from v1 to v2, and changes back to v1, while serving CR write requests in each stage. We cannot just update the storage version to v1 and unblock all the write requests, because the storage migrator won't be able to tell that v2 data exists in etcd. Therefore, there can be more than one update for a UID in the queue at some point.

Reply:
A possible optimization is to merge consecutive updates triggered by watch event handlers. I added a TODO